tools/ports/__init__.py

```diff
@@ -394,7 +394,9 @@ def dfs(node):
 
 def resolve_dependencies(port_set, settings):
   def add_deps(node):
-    node.process_dependencies(settings)
+    port_dependencies = node.process_dependencies(settings)
+    if port_dependencies is not None:
```

Comment:
I think just `if port_dependencies:` is fine here.

Reply:
If you look at sdl2_image, it does two things during `process_dependencies`.

1. Update settings
2. Update its own `.deps` list

I have two siggestions here:

1. Can we stick to the mechanism of updated `.deps` during `process_dependencies` rather than returning the deps?
2. I think `resolve_dependencies` is supposed to add the dependencies to `port_set`,  I think it would then be up to the calling code to add them to `ports_needed`.

Reply:
I am not sure I fully understand. Are you suggesting that I should do this in my code (and no change in emscripten at all):

```python
# imgui.py port
def process_dependencies(settings):
    ports_needed.add('contrib.glfw3')
```

If I can get a hold on `ports_needed` then clearly that could work.

Reply:
No, I believe you your dependencies are fixed/static then you shouldn't need to implement `process_dependencies` at all. 

It should be enough to just write `deps = ['contrib.glfw3']`

`process_dependencies` should only be needed for ports that have dyanmic/conditional dependencies.

Reply:
So how do I fix the link issue then?

With "old" port there is no link issue. With new ports there is one. I don't think that makes sense.

```text
# this is fine and will automatically link with sdl2 even if not explicitly added to the command line
emcc --use-port=sdl2_mixer test/hello_world.c -o /tmp/out.js
```

Why would it work for "old" port but not new ones?

Reply:
You said:

> I think resolve_dependencies is supposed to add the dependencies to port_set, I think it would then be up to the calling code to add them to ports_needed.

I don't understand what is the calling code you are referring to that would "add them to ports_needed"

Reply:
I think the problem is that we don't add the deps to the new `ports_needed` global.

We have a `if port.needed(settings):` check on line 500 which we should perhaps delete.

Alternatively you could make `port.needed(settings)` return true if you add  `ports_needed.add(dep)` right after `port_set.add(deps)` perhaps?

I think perhaps we should do both of the above since the check on line 500 seems redundant, but it would also be nice if `.needed()` returned the correct thing for contrib deps

Reply:
Ok I will investigate further

In regard to the statement 

> process_dependencies should only be needed for ports that have dyanamic/conditional dependencies.

I believe it is not entirely accurate. You can check [vorbis](https://github.com/emscripten-core/emscripten/blob/main/tools/ports/vorbis.py) port:

```python
def process_dependencies(settings):
  settings.USE_OGG = 1
```

There is nothing dynamic here. This function lets you write `emcc -sUSE_VORBIS=1 ...` without having to write `emcc -sUSE_OGG=1 -sUSE_VORBIS=1 ...` for linking purposes...

Reply:
Ah yes, I see what you mean.  I guess we have a lot of duplication between `process_dependencies` and `deps`. 

I would hope that we could simply rely on `deps` along rather then duplication between them. e.g.:

```
deps = ['freetype', 'sdl2', 'harfbuzz']  

...

def process_dependencies(settings):                                                                  
  settings.USE_SDL = 2                                                                               
  settings.USE_FREETYPE = 1                                                                          
  settings.USE_HARFBUZZ = 1    
```

At least for contrib ports it should be easy enough to just use `deps` I think.. since you don't have any settings that you need to set.